### DISCUSSION TITLE
G-Helper keeps changing fancurve by +-1 degree

### DISCUSSION THREAD

**Discussion by Candelus** [seerge/g-helper#1744] [Q&A] [answered]
I'm a little bit OCD so having the fancurve not obeying what I set it to drives me crazy. This wasn't a thing with Armoury Crate custom fan curve, any suggestions to fix it?

<img width="507" alt="2023-12-15 23_40_32" src="https://github.com/seerge/g-helper/assets/153918815/5097bdd0-cc2c-4d4c-9798-42cc647973ba">

**Comment by seerge** [comment]
> [How does G-helper control my fan speeds?](https://github.com/seerge/g-helper?tab=readme-ov-file#how-does-g-helper-control-my-fan-speeds)
> It doesn't. Your firmware / BIOS controls them in real-time. Armoury also doesn't control fans in real time anyhow.

> What G-helper can do - is (optionally) set a custom fan profile to the current performance mode consisting of 8 pairs of temperature + fan speed % via the same endpoint Armoury seems to use. How it will be interpreted - is still up to the firmware.

**Reply by Candelus** [reply]
So it is the firmware that fluctuates and G-Helper shows the result, is that correct?

**Reply by seerge** [reply]
@Candelus if you mean fan speed you see on main app window, yes - that what bios reports and app just shows it

**Reply by Candelus** [reply]
@seerge Oh no, I mean on the Fans and Power windows like on the screenshot. Everytime I set it carefully to like 50 55 60 65 70 75 it would eventually fluctuate to 49, 56, 61 etc. I'm making the assumption based on your writings that we set it how we want, and how it will look like will be interpreted by the firmware -> the 1 degree fluctuation after setting the fan curve is what the firmware interpreted which G-Helper shows the result returned by the firmware.  Is this correct? 

> What G-helper can do - is (optionally) set a custom fan profile to the current performance mode consisting of 8 pairs of temperature + fan speed % via the same endpoint Armoury seems to use. How it will be interpreted - is still up to the firmware.

**Reply by seerge** [reply, marked as answer]
@Candelus i see what you mean. It's connected with rounding of a (floating) dot coordinates on a chart to an integer value of temperature.

Every time you would click Apply it saves actual curve into config, and redraws a chart based on already rounded values, which would make points jump slightly :)

I have adjusted rounding a bit, so it will always show correct temperature in a tooltip. So temps "won't" change 

Try this build

[GHelper.zip](https://github.com/seerge/g-helper/files/13695674/GHelper.zip)

**Reply by Candelus** [reply]
@seerge thank you so much because this seems to work! It's a bit weird that after clicking apply the dot still jumps a bit, yet the temp remains the same as I when I set it, but as long as it is the number I want it's fine with me.

**Reply by seerge** [reply]
@Candelus yes, it jumps down (left) a bit, as chart lets you set dot to something like 93.5C, but when it's being saved / applied it becomes 93C
